File: daily-programmer/1-easy-input.hs
{-

    https://www.reddit.com/r/dailyprogrammer/comments/pih8x/easy_challenge_1/

    [easy] challenge #1

    create a program that will ask the users name, age, and reddit username.
    have it tell them the information back, in the format:

        your name is (blank), you are (blank) years old, and your username is (blank)

    for extra credit, have the program log this information in a file to be accessed later.

-}

import System.IO (hFlush, stdout)

main :: IO()
main = output . formMsg =<< input
    where output str = mapM_ ($ str) [putStr, appendFile "tmp/log.txt"]

input :: IO (String, String, String)
input = do
    name <- prompt "your name"
    age <- prompt "your age"
    username <- prompt "your username"
    return (name, age, username)

formMsg :: (String, String, String) -> String
formMsg (name, age, username) =
    concat
        [ "your name is ", name, ", "
        , "you are ", age, " years old, "
        , "and your username is ", username
        , "\n"
        ]

-- UTILS

prompt :: String -> IO String
prompt msg = do
    putStr $ "input " ++ msg ++ " > "
    hFlush stdout
    getLine

File: daily-programmer/1-easy-input.c
/*

    https://www.reddit.com/r/dailyprogrammer/comments/pih8x/easy_challenge_1/

    [easy] challenge #1

    create a program that will ask the users name, age, and reddit username.
    have it tell them the information back, in the format:

        your name is (blank), you are (blank) years old, and your username is (blank)

    for extra credit, have the program log this information in a file to be accessed later.

*/

#include <stdio.h>
#include <string.h>

#define LENGTH 255

int main(
    void
) {
    const char format[] = "your name is %s, you are %s years old, and your username is %s\n";
    char name[LENGTH], age[LENGTH], username[LENGTH];
    FILE *log;

    printf("input your name > ");
    gets(name);
    printf("input your age > ");
    gets(age);
    printf("input your username > ");
    gets(username);

    printf(format, name, age, username);

    log = fopen("tmp/log.txt", "a");
    if (!log) {
        fprintf(stderr, "could not open file log.txt\n");
        return -1;
    }
    fprintf(log, format, name, age, username);
    fclose(log);
}
